network/src/validator_network/network_builder.rs

```diff
@@ -414,14 +414,22 @@ impl NetworkBuilder {
             .take()
             .expect("Authentication Mode not set");
 
-        let (key, maybe_trusted_peers) = match authentication_mode {
-            AuthenticationMode::ServerOnly(key) => (key, None),
-            AuthenticationMode::Mutual(key) => (key, Some(self.trusted_peers.clone())),
+        let (key, maybe_trusted_peers, peer_id) = match authentication_mode {
+            AuthenticationMode::ServerOnly(key) => {
+                let public_key = key.public_key();
+                // TODO: should I check that self.peer_id is not set then?
+                let peer_id = PeerId::from_identity_public_key(public_key);
+                (key, None, peer_id)
+            }
```

Comment:
This isn't accurate and perhaps was part of my hangup in the previous PR. The upstream peers ID may be in a trusted peers from upstream. If I am the client, my peer id is derived this way, but my peer id may not be derived.

Reply:
what previous PR btw : o ?

not sure I understand your comment. Why would you set a peer_id if you're in the full node network?

Reply:
ok I implemented the correct behavior now :)

```rust
        let (key, maybe_trusted_peers, peer_id) = match authentication_mode {
            // validator-operated full node
            AuthenticationMode::ServerOnly(key) if self.peer_id == PeerId::default() => {
                let public_key = key.public_key();
                let peer_id = PeerId::from_identity_public_key(public_key);
                (key, None, peer_id)
            }
            // full node
            AuthenticationMode::ServerOnly(key) => (key, None, self.peer_id),
            // validator
            AuthenticationMode::Mutual(key) => {
                (key, Some(self.trusted_peers.clone()), self.peer_id)
            }
        };
```

Reply:
cc @phlip9 I'm wondering why we don't just make the distinction between the three types of networks we have?

There's also an argument to make AuthenticationMode hold the peer_id as well.